data-prepper-plugins/saas-source-plugins/saas-crawler/src/test/java/org/opensearch/dataprepper/plugins/source/saas_crawler/coordination/scheduler/LeaderSchedulerTest.java

```diff
@@ -0,0 +1,110 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testLeaderPartitionsCreation(boolean initializationState) throws InterruptedException {
+        LeaderScheduler leaderScheduler = new LeaderScheduler(coordinator, saasSourcePlugin, crawler);
+        LeaderPartition leaderPartition = new LeaderPartition();
+        leaderPartition.getProgressState().get().setInitialized(initializationState);
+        leaderPartition.getProgressState().get().setLastPollTime(0L);
+        given(coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE)).willReturn(Optional.of(leaderPartition));
+        doThrow(RuntimeException.class).when(coordinator).saveProgressStateForPartition(any(LeaderPartition.class), any(Duration.class));
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(leaderScheduler);
+
+        Thread.sleep(100);
+        executorService.shutdownNow();
+
+        // Check if crawler was invoked and updated leader lease renewal time
+        verify(crawler, times(1)).crawl(0L, coordinator);
+        verify(coordinator).saveProgressStateForPartition(eq(leaderPartition), any(Duration.class));
+
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void testExceptionWhileAcquiringLeaderPartition(boolean initializationState) throws InterruptedException {
+        LeaderScheduler leaderScheduler = new LeaderScheduler(coordinator, saasSourcePlugin, crawler);
+        LeaderPartition leaderPartition = new LeaderPartition();
+        leaderPartition.getProgressState().get().setInitialized(initializationState);
+        leaderPartition.getProgressState().get().setLastPollTime(0L);
+        given(coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE)).willThrow(RuntimeException.class);
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(leaderScheduler);
+
+        Thread.sleep(100);
+        executorService.shutdownNow();
+
+        // Check if crawler was invoked and updated leader lease renewal time
+        verifyNoInteractions(crawler);
+    }
+
+    @Test
+    void testWhileLoopRunnningAfterTheSleep() throws InterruptedException {
+        LeaderScheduler leaderScheduler = new LeaderScheduler(coordinator, saasSourcePlugin, crawler);
+        LeaderPartition leaderPartition = new LeaderPartition();
+        leaderPartition.getProgressState().get().setInitialized(false);
+        leaderPartition.getProgressState().get().setLastPollTime(0L);
+        given(coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE)).willReturn(Optional.of(leaderPartition));
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(leaderScheduler);
+
+        //Wait for more than a minute as the default while loop wait time in leader scheduler is 1 minute
+        Thread.sleep(70000);
```

Comment:
This is a very long sleep. See comments elsewhere.

This one definitely needs some improvement.

Reply:
Make the leader scheduler schedule time configurable and configure it to smaller values in tests.

Reply:
Thank you @kkondaka for the input. I made the scheduler time configurable and able to reduce the wait time to 100 milli seconds for this test